packages/app-builder/src/components/Scenario/Workflow/WorkflowProvider.tsx

```diff
@@ -0,0 +1,251 @@
+            nodes,
+            edges,
+          });
+        },
+      },
+    })),
+  );
+
+  const dataValue = React.useMemo(
+    () => ({
+      inboxes: data.inboxes,
+      scenarios: data.scenarios,
+    }),
+    [data.inboxes, data.scenarios],
+  );
+
+  return (
+    <WorkflowDataContext.Provider value={dataValue}>
+      <WorkflowStoreContext.Provider value={store}>
+        {children}
+      </WorkflowStoreContext.Provider>
+    </WorkflowDataContext.Provider>
+  );
+}
+
+function clearSelection<T extends { selected?: boolean }>(items: T[]): T[] {
+  return items.map((node) => ({ ...node, selected: false }));
+}
+
+function createEmptyNode(): Node<EmptyNodeData> {
+  return {
+    id: nanoid(6),
+    type: 'empty_node',
+    data: {
+      type: 'empty-node',
+    },
+    position: { x: 0, y: 0 },
+  };
+}
+
+// Layout settings to be used by Dagre and addEmptyNode: define the distance between nodes
+export const nodesep = 100;
+
+function shouldCreateEmptyNode(
+  nodes: Node<NodeData>[],
+  edges: Edge[],
+): boolean {
+  if (nodes.length === 0) return true;
+  if (nodes.some((node) => isTriggerData(node.data))) return false;
+
+  const emptyNodes = nodes.filter((node) => node.type === 'empty_node');
+  for (const emptyNode of emptyNodes) {
+    const connectedEdges = getConnectedEdges([emptyNode], edges);
+    if (connectedEdges.length === 0) return false;
+  }
+  return true;
+}
+
+function createInitialState(): WorkflowState {
+  const emptyNode = createEmptyNode();
+  emptyNode.selected = true;
+  return {
+    nodes: [emptyNode],
+    edges: [],
+  };
+}
+
+function useWorkflowStore<Out>(selector: (state: WorkflowStore) => Out) {
+  const store = WorkflowStoreContext.useValue();
+  return useStore(store, selector);
+}
+
+export function useNodes() {
+  return useWorkflowStore((state) => state.nodes);
+}
+
+export function useEdges() {
+  return useWorkflowStore((state) => state.edges);
+}
+
+export function useIsSourceConnectable({ nodeId }: { nodeId: string }) {
+  const edges = useWorkflowStore((state) => state.edges);
+
+  return React.useMemo(
+    () => !edges.some((edge) => edge.source === nodeId),
+    [edges, nodeId],
+  );
+}
+
+export function useSelectedNodes() {
+  const nodes = useNodes();
+  return nodes.filter((node) => node.selected === true);
+}
+
+/**
+ * Returns the type of node that can be created in place of the given node.
+ * If a trigger already exists, it can only be replaced by an action.
+ * If an incoming edge exists, the node can only be an action.
+ * Otherwise, the node can only be a trigger.
+ */
+export function useCreateNodeType(nodeId: string) {
+  const edges = useEdges();
+  const nodes = useNodes();
+  if (edges.some((edge) => edge.target === nodeId)) {
+    return 'action';
+  }
+  if (nodes.some((node) => isTriggerData(node.data))) {
+    return 'action';
+  }
+  return 'trigger';
+}
+
+export function useWorkflowActions() {
+  return useWorkflowStore((state) => state.actions);
+}
```

Comment:
This file is were the internal state logic is implemented :
- internal state + actions (= zustand)
- external data required to display contextual information like scenario, inboxes names (= Classic React Context)